### PR TITLE
Migrate add phone number dialog to Mobius - Stabilize

### PR DESCRIPTION
**Pull request by msasikanth**
https://app.clubhouse.io/simpledotorg/story/347/migrate-add-phone-number-dialog-to-mobius

### CODE REVIEW

**Comment by msasikanth** [comment]
Marking it as `WIP` in order to fix parameter tests.

**Comment by msasikanth** [comment]
I have rebased commits, please check out the branch when reviewing

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt [hunk 1]
```diff
@@ -118,11 +112,14 @@ class AddPhoneNumberDialogControllerTest {
...
-  private fun setupController() {
+  private fun setupController(newNumber: String) {
     val controller = AddPhoneNumberDialogController(repository, validator, FakeUuidGenerator.fixed(generatedPhoneUuid))
 
     controllerSubscription = uiEvents
         .compose(controller)
         .subscribe { uiChange -> uiChange(dialog) }
+
+    uiEvents.onNext(AddPhoneNumberDialogCreated(patientUuid))
+    uiEvents.onNext(AddPhoneNumberSaveClicked(newNumber))
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:129]: Can we have this particular event done separately in each test? We should only move emission of the `ScreenCreated` event since that is a part of screen initialization and something that is supposed to be triggered for every test.

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt [hunk 2]
```diff
@@ -71,6 +72,7 @@ class AddPhoneNumberDialogControllerTest {
         numberDetails = numberDetails,
         active = true
     )
+    verifyNoMoreInteractions(repository)
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:79]: Can we also verify no interactions on the `dialog` here?

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt [hunk 3]
```diff
@@ -90,6 +92,7 @@ class AddPhoneNumberDialogControllerTest {
     setupController(newNumber)
 
     verify(repository, never()).createPhoneNumberForPatient(any(), any(), any(), any())
+    verifyNoMoreInteractions(repository)
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:99]: Can we also verify no interactions on the `dialog` here?

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt [hunk 4]
```diff
@@ -79,13 +80,10 @@ class AddPhoneNumberDialogControllerTest {
...
   @Test
-  @Parameters(method = "validation errors")
-  fun `when save is clicked, the number should not be saved if it's invalid`(
-      validationError: Result
-  ) {
+  fun `when save is clicked, the number should not be saved if it's invalid`() {
     // given
     val newNumber = "123"
-    whenever(validator.validate(newNumber, type = LANDLINE_OR_MOBILE)).thenReturn(validationError)
+    whenever(validator.validate(newNumber, type = LANDLINE_OR_MOBILE)).thenReturn(validationErrors.random())
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:86]: When we remove parameterized tests, we should be adding a separate test for each one of the removed parameter combinations instead of using a random value. Can we do that in this commit?

Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:86]: Can we use a real instance of `PhoneNumberValidator` here? See step 3 in [`Stabilize`](https://github.com/simpledotorg/simple-android/blob/master/doc/mobius/migrating-to-mobius.md#steps).

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt [hunk 5]
```diff
@@ -118,11 +112,14 @@ class AddPhoneNumberDialogControllerTest {
...
 
-  private fun setupController() {
+  private fun setupController(newNumber: String) {
     val controller = AddPhoneNumberDialogController(repository, validator, FakeUuidGenerator.fixed(generatedPhoneUuid))
 
     controllerSubscription = uiEvents
         .compose(controller)
         .subscribe { uiChange -> uiChange(dialog) }
+
+    uiEvents.onNext(AddPhoneNumberDialogCreated(patientUuid))
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:128]: One more thing I see that is missing in this PR is changing the controller to accept the parameters via assisted injection instead of a custom `ScreenCreated` event (See step 7 of [`Stabilize`](https://github.com/simpledotorg/simple-android/blob/master/doc/mobius/migrating-to-mobius.md#steps)).

Was this missed or we going to do it in another PR?

Comment (msasikanth) [reply app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:128]: Ah right, I missed it. Will fix it

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt (hunk 4 above)
Comment (msasikanth) [reply app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:86]: I initially went with that approach and it seemed like a lot of code duplication. So using random felt like a good option, but will change back to having different tests.

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt (hunk 1 above)
Comment (msasikanth) [reply app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:129]: You're talking about `AddPhoneNumberSaveClicked` event right?

Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:129]: Yes.

File: app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt (hunk 4 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/summary/addphone/AddPhoneNumberDialogControllerTest.kt:86]: The problem with this is that it adds non-determinism to the tests, which makes the test suite unreliable. Code duplication is encouraged in tests to keep them isolated and focused.
